### PR TITLE
Add recipe side-hustle

### PR DESCRIPTION
**Pull request by rnkn**
### Brief summary of what the package does

Navigate Imenu as a list in a side window. Press RET to go to an item, SPC to show. Items with child items can be collapsed/expanded.

This package achieves a very similar thing as [Imenu-list](https://github.com/bmag/imenu-list) but takes a different approach, permitting navigation of multiple buffers, doesn't require a global minor mode, or timers, or hide-show, and uses dedicated side-windows

### Direct link to the package repository

https://github.com/rnkn/side-hustle

### Your association with the package

Author

### Relevant communications with the upstream package maintainer

**None needed**

### Checklist

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses)
- [x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback

n.b. I include a convenience alias `toggle-side-hustle` for command `side-hustle-toggle`.

- [x] My elisp byte-compiles cleanly
- [x] `M-x checkdoc` is happy with my docstrings
- [x] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [ ] I have confirmed some of these without doing them

### CODE REVIEW

**Comment by riscy** [comment]
> n.b. I include a convenience alias toggle-side-hustle for command side-hustle-toggle.

I would push back against this if package-lint is complaining (package-lint does have some exceptions built in).  In this case there is a [package called toggle](https://melpa.org/#/toggle).  You _could_ suggest users do such a thing in your `;;; Commentary`.  Here's a canned explanation for why: It may feel inconvenient, but a simple and fair way to prevent namespace collisions across MELPA (and hopefully elsewhere) is to ask that the package, variable names, and function names all use the same prefix.  In very rare cases I would make exceptions to this rule (e.g. backward compatibility with a package predating this rule) -- and `package-lint` has some of these exceptions built into it.

Beyond that this looks good - so thanks for paying close attention to the checklist.  I have a couple other minor issues, one of which might disappear if the alias is addressed:

- side-hustle.el#L327: It's safer to sharp-quote function names; use `#'`
- side-hustle.el#L276: Could this point/point-max comparison use `eobp`?

**Comment by rnkn** [comment]
Thanks for the thorough vetting!

I think the `toggle` package is at fault here, given one need only open an emacs -Q session and `M-x toggle-[TAB]` to see that "toggle" is a long-established generic prefix. I'd prefer to keep this as is (but I will add the forgotten sharp quote, thanks!).

You're right that a while loop like that could also be `(while (not (eobp)) ...)`; this is just a personal preference that I like my movement loops to take the form `(while (< (point) ENDPOINT) ...)`. I find this more idiomatic and easier for me when I come back to the code.

**Comment by riscy** [comment]
> I find this more idiomatic and easier for me when I come back to the code.

Yeah, that's completely reasonable.  That lint may be more noise than signal. :)

> "toggle" is a long-established generic prefix. I'd prefer to keep this as is 

I think you have a good point (especially about the `toggle` package sitting in a crowded namespace and may have been added before MELPA's present day use of package-lint to nail down prefix rules and exceptions).  And while toggle-prefixed functions do appear throughout Emacs, I have one small counterexample:  `toggle-read-only` is marked obsolete in favor of `read-only-mode`.

But let me ping @purcell on this if he gets a chance to look.  Perhaps this is a prefix exception that could be built into package-lint, and then I would merge without even thinking about it.  I'm giving you a hard time because it's easier (overall) just to lean on the tools than to negotiate exception cases.

**Comment by rnkn** [comment]
No problemo. Yeah I was going to suggest a `toggle-PREFIX-*` exception, similar to the `define-*` for macros.

Just looking at the history for `toggle-read-only` I think marking of that as obsolete may have been more to do with it being intended as interactive only but proliferating in elisp.

**Comment by alphapapa** [comment]
This looks like a nice improvement over `imenu-list`.  What do you think about upstreaming it into Emacs?  I certainly wish that Imenu had something like this built-in; it seems like a shame that it doesn't.

**Comment by rnkn** [comment]
@alphapapa Thanks, but that would be my worst nightmare. Dealing with emacs-devel is seriously one of the most painful aspects of my life.

**Comment by alphapapa** [comment]
That's too bad.  I read the list from time to time, and I thought you were a regular contributor.  Seems like a pretty straightforward feature to add.

Anyway, if I might make one final suggestion: while I really like the name `side-hustle`, since it doesn't have `imenu` in the name, I think it would be destined for obscurity among imenu users.

**Comment by rnkn** [comment]
@alphapapa yeah but then "olivetti" is pretty obscure and that's super popular.

**Comment by alphapapa** [comment]
Yes, but olivetti is its own mode, not an enhancement for a specific, built-in package.  

I think that users who are looking for imenu enhancements would be more likely to discover a package named `imenu-list` than one named `side-hustle`, which would lead to them using the older, inferior imenu-list rather than the newer, superior side-hustle.  

I've seen this happen over and over with other Emacs packages: the older package is more discoverable (and already mentioned all over the Internet), so new users find and use it rather than the newer, better alternative.  With nearly 5,000 packages in MELPA alone, discoverability is a real obstacle to adoption of newer packages.  Names really do matter.

**Comment by rnkn** [comment]
@alphapapa all valid points, but popularity isn't really something I'm seeking. I like the name and that's the main thing for me. If someone does `C-s imenu` in the packages window they'll still see "Hustle through Imenu in a side window".

I wouldn't necessarily say this is superior to Imenu-List, it just takes a different approach.

**Comment by riscy** [comment]
Hi @rnkn, I think the name is good if you like it (although I often agree with @alphapapa's sentiment that our elisphere tends to have a very large and varied vocabulary!) -- and the code looks great.

I opened an issue which you can see linked above; I think there is good rationale there for not having package-lint except such prefixes.  I think it would be best in this case if you just suggested `(defalias 'toggle-side-hustle #'side-hustle-toggle)` to users in your `;;; Commentary` if they want to use such an alias.  I think the simpler and more consistent our treatment of this prefix-as-a-namespace constraint is, the better for everyone.

**Comment by rnkn** [comment]
Cool. I just commented it out. If anyone really wants the `toggle-` command they can find it in the code and uncomment.

**Comment by riscy** [comment]
Great - thank you for accommodating.

**Comment by alphapapa** [comment]
@rnkn If I may, I'd like to suggest making a clarification in the package's docs and description.  It currently says, "spawns a side window linked to the current buffer, which allows working with multiple buffers simultaneously," which I interpreted to mean that, after opening the side-hustle window, when I then switch to another buffer (B) in the window which was selected when I opened the side-hustle window, the side-hustle window would automatically update to show the contents of the newly switched-to buffer (B).  But what seems to happen is that the side-hustle window continues showing the contents of the buffer (A) that was shown when I opened the side-hustle window.

In other words, one of the appeals of this package, to me, was that I could open a single side-hustle window, and it would "follow" me to other source files, always showing me the imenu contents for what I was looking at.  If this isn't what it does, then I must be missing something.  :)  OTOH, maybe I'm doing something wrong?

Thanks.

**Comment by rnkn** [comment]
@alphapapa ah right, no that's what Imenu-List does, whereas my intent here was to take the other approach, i.e. you can switch around the buffer in the main window/s, then use the Side Hustle side window to jump back to where you were.

**Comment by alphapapa** [comment]
Yeah, I had it backwards: I thought side-hustle did what imenu-list does but better.  I can see how both could be useful.  Thanks.

**Comment by rnkn** [comment]
Nah I try to avoid writing something motivated by NIH; I think that might make the other author feel bad. But I appreciate you making me aware of the ambiguity of the README, thanks :)
